nav2_theta_star_planner/include/nav2_theta_star_planner/theta_star.hpp

```diff
@@ -0,0 +1,244 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_THETA_STAR_PLANNER__THETA_STAR_HPP_
+#define NAV2_THETA_STAR_PLANNER__THETA_STAR_HPP_
+
+#include <cmath>
+#include <chrono>
+#include <vector>
+#include <queue>
+#include <algorithm>
+#include "rclcpp/rclcpp.hpp"
+#include "nav2_costmap_2d/costmap_2d_ros.hpp"
+
+const double INF_COST = DBL_MAX;
+const int LETHAL_COST = 252;
+
+struct coordsM
+{
+  int x, y;
+};
+
+struct coordsW
+{
+  double x, y;
+};
+
+struct pos
+{
+  int pos_id;
+  double f;
+};
+
+struct tree_node
+{
+  int x, y;
+  double g = INF_COST;
+  double h = INF_COST;
+  int parent_id;
+  bool is_in_queue = false;
+  double f = INF_COST;
+};
+
+struct comp
+{
+  bool operator()(pos & p1, pos & p2)
+  {
+    return (p1.f) > (p2.f);
+  }
+};
+
+/// TODO(ANSHU-MAN567): try using a pointer to the tree_node vector instead of directly using it
+
+namespace theta_star
+{
+class ThetaStar
+{
+public:
+  coordsM src_{}, dst_{};
+  nav2_costmap_2d::Costmap2D * costmap_{};
+
+  /// weight for the costmap traversal cost
+  double w_traversal_cost_;
+  /// weight for the euclidean distance cost (as of now used for calculations of g_cost)
+  double w_euc_cost_;
+  /// weight for the heuristic cost (for h_cost calculcations)
+  double w_heuristic_cost_;
+
+  int how_many_corners_;
+  int size_x_, size_y_;
+
+  ThetaStar();
+
+  /**
+   * @brief the function that iteratively searches upon the nodes in the queue (open list) until the
+   *            current node is the goal pose or if size of queue > 0
+   * @param raw_path is used to return the path obtained on exectuing the algorithm
+   * @return true if a path is found, false if no path is found between the start and goal pose
+   */
+  bool generatePath(std::vector<coordsW> & raw_path);
+
+  bool isSafe(const int & cx, const int & cy) const
+  {
+    return costmap_->getCost(cx, cy) < LETHAL_COST;
+  }
+
+protected:
+  /// for the coordinates (x,y), we store at node_position_[size_x_ * y + x],
+  /// the index at which the data of the node is present in nodes_data_
+  std::vector<int> node_position_;
+
+  /// the vector nodes_data_ stores the coordinates, costs and index of the parent node,
+  /// and whether or not the node is present in queue_
+  std::vector<tree_node> nodes_data_;
+
+  /// this is the priority queue (open_list) to select the next node for expansion
+  std::priority_queue<pos, std::vector<pos>, comp> queue_;
+
+  /// it is a counter like variable used to generate consecutive indices
+  /// such that the data for all the nodes (in open and closed lists) could be stored
+  /// consecutively in nodes_data_
+  int index_generated_;
+
+  const coordsM moves[8] = {{0, 1},
+    {0, -1},
+    {1, 0},
+    {-1, 0},
+    {1, -1},
+    {-1, 1},
+    {1, 1},
+    {-1, -1}};
+
+  tree_node * curr_node = new tree_node;
```

Comment:
Won't this leak memory? Why is this set in the header?

Reply:
Well it is supposed to prevent calling the curr_node multiple times in the code. I have used it in a bunch of different places and thought that rather than referencing it in all the functions I need it in, I could make a variable that points the current node

Reply:
Sure, but why are you allocating here? You should remove `new tree_node;`